CartoDB/cartodb
# Adds export image

This PR adds the ability to export maps as images.

- This feature already exists in the Editor, but wasn't available for Builder.
- Depends on https://github.com/CartoDB/cartodb.js/issues/1595
- Depends on the `static-maps-layers-filter` of Windshaft.

![export](https://cloud.githubusercontent.com/assets/4933/24053547/ac3958f8-0b39-11e7-99e8-8af1fea40e99.gif)

### Characteristics:

- Users can change the width & height of a rectangle that will be used as a reference to generate the image.
- Images can be exported as JPG or PNG files.
- The images are limited to 8192 x 8192 pixels.
- Image resolution is set to 72 DPI.
- JPEG quality is 85%.
- The name of the exported image will be: *Name of the map*.{jpg|png}.

### Future enhancements

- Google Maps basemaps won't appear right now. This will be handled in a different ticket.
- We'll implement the inclusion of the CARTO logo and the map attribution in a future ticket.
- Before we implement the export of legends, they will be hidden from the view.

### Tracking of this feature

This feature contains certain classes to track its use:

`.track-ImageExport.track-typeJPG`
`.track-ImageExport.track-typePNG`
`.track-ImageExport.track-inputWidth`
`.track-ImageExport.track-inputHeight`
`.track-ImageExport.track-export`
`.track-ImageExport.track-nodeTopLeft`
`.track-ImageExport.track-nodeTopCenter`
`.track-ImageExport.track-nodeTopRight`
`.track-ImageExport.track-nodeMiddleLeft`
`.track-ImageExport.track-nodeMiddleCenter`
`.track-ImageExport.track-nodeBottomLeft`
`.track-ImageExport.track-nodeBottomCenter`
`.track-ImageExport.track-nodeBottomRight`

### Extra stuff

Beside the main feature, this PR adds a couple of changes to the existing codebase:

- Adds the ability to add classes in the text and radio form fields.
- Adds specs for the text and radio form fields.
- Adds the ability to increase/decrease the value of the text fields using the up & down arrows (using the shift key makes the increment/decrement bigger)
- Adds the ability to add clases to the `StackLayoutViews`.

### Feature flag

This feature is protected by the `static-maps` feature flag.

### How to test this

Enable the `static-maps` feature flag and then simply create a map (with or without geometries) and export it. Change the basemap and export it again (remember that at this moment we don't export Google Maps basemaps).

This PR requires having [this cartodb.js PR](https://github.com/CartoDB/cartodb.js/pull/1598). If you are doing acceptance and the PR is not already merged, point the cartodb.js dependency to the #static-images branch.

See original ticket here: #10822
